core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -437,6 +484,22 @@ private[spark] class SparkSubmit extends Logging {
       }.orNull
     }
 
+    // At this point, we have attempted to download all remote resources.
+    // Now we try to resolve the main class if our primary resource is a JAR.
+    // First, try the localPrimaryResource.
+    if (args.mainClass == null) {
+      args.mainClass = resolveMainClassIfNeeded(args, hadoopConf, localPrimaryResource)
+    }
+    // If we can't get it from localPrimaryResource, use whatever the original argument was.
```

Comment:
In what case is `localPrimaryResource` not a local copy of `args.primaryResource`?

If seems to me that what you want is:

```
if (mainClass == null && not python && not R) {
  mainClass = readMainClass(Option(localPrimaryResource).getOrElse(args.primaryResource), hadoopConf)
  if (mainClass == null) yellAtUser()
}
```

Reply:
> In what case is `localPrimaryResource` not a local copy of `args.primaryResource`?

I'm not quite sure actually, but right above this code is a `clusterManager == YARN` check where it seems to be using `args.primaryResource` instead of `localPrimaryResource`, so I just tried to check both variables?